### PR TITLE
feat: integrate multiplexing

### PR DESCRIPTION
**Pull request by mattsse**
refactor some multiplex internals and add test for test protocol.

### CODE REVIEW

File: crates/net/eth-wire/src/test_utils.rs [hunk 1]
```diff
@@ -55,3 +55,106 @@ pub async fn connect_passthrough(
...
+        Ping,
+        Pong,
+    }
+
+    /// An `test` protocol message, containing a message ID and payload.
+    #[derive(Clone, Debug, PartialEq, Eq)]
+    pub struct TestProtoMessage {
+        pub message_type: TestProtoMessageId,
+        pub message: TesProtoMessage,
+    }
```
Comment (gakonst) [inline crates/net/eth-wire/src/test_utils.rs:92]: `Test`

**Review by Rjected — COMMENTED** [review]
I like the tests and updated inner type, just left some notes on extracting some parts of the poll loop to methods

File: crates/net/network/src/session/conn.rs [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+/// Either a [`EthPeerConnection`] or an [`EthSatelliteConnection`].
+// This type is boxed because the underlying stream is ~6KB,
+// mostly coming from `P2PStream`'s `snap::Encoder` (2072), and `ECIESStream` (3600).
+#[derive(Debug)]
+pub enum EthRlpxConnection {
+    /// A That only supports the ETH protocol.
+    EthOnly(Box<EthPeerConnection>),
+    /// A connection that supports the ETH protocol and __at least one other__ RLPx protocol.
+    Satellite(Box<EthSatelliteConnection>),
+}
```
Comment (emhane) [inline crates/net/network/src/session/conn.rs:36]: is starting a `SatelliteStream` that much overhead when no extra protocols are shared that this enum is necessary?

Comment (mattsse) [reply crates/net/network/src/session/conn.rs:36]: possibly, because this would involve routing through a channel,

**Review by Rjected — APPROVED** [review]
ok, this looks good, super excited to see multiple caps in practice

**Comment by mattsse** [comment]
After thinking a bit longer about this,
I'm going with this satellite approach over the `MuxDemuxer` because
* this still allows us to run a regular wrapped `EthStream<TcpStream>` most of the time (if no custom handlers installed and no matching additional caps) and we don't have any overhead routing through channels
* the SatelitteStream is convenient because it hides polling the other cap streams internally so the interface can be unified over the `enum EthRlpxConnection`

**Comment by emhane** [comment]
> After thinking a bit longer about this, I'm going with this satellite approach over the `MuxDemuxer` because
> 
> * this still allows us to run a regular wrapped `EthStream<TcpStream>` most of the time (if no custom handlers installed and no matching additional caps) and we don't have any overhead routing through channels
`MuxDemuxer` doesn't route the main stream through channels
> * the SatelitteStream is convenient because it hides polling the other cap streams internally so the interface can be unified over the `enum EthRlpxConnection`
polling the main stream that owns the `MuxDemuxer` polls the other cap streams

**Pull request merged**